src/librustc_target/abi/mod.rs

```diff
@@ -897,25 +920,31 @@ pub enum DiscriminantKind {
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
 pub struct Niche {
     pub offset: Size,
-    pub scalar: Scalar,
+    pub scalar: Scalar<Integer>,
 }
 
 impl Niche {
-    pub fn from_scalar<C: HasDataLayout>(cx: &C, offset: Size, scalar: Scalar) -> Option<Self> {
-        let niche = Niche {
-            offset,
-            scalar,
-        };
-        if niche.available(cx) > 0 {
+    pub fn from_int(offset: Size, scalar: Scalar<Integer>) -> Option<Self> {
+        let niche = Niche { offset, scalar };
+        if niche.available() > 0 {
             Some(niche)
         } else {
             None
         }
     }
 
-    pub fn available<C: HasDataLayout>(&self, cx: &C) -> u128 {
+    pub fn from_scalar<C: HasDataLayout>(cx: &C, offset: Size, scalar: Scalar) -> Option<Self> {
+        let Scalar { value, valid_range } = scalar;
+        let scalar = Scalar {
+            value: Integer::for_size(value.size(cx)).unwrap(),
```

Comment:
Why does this size call take CX and the others don't?

Reply:
Compare `Integer::size` and `Primitive::size`. I even added a `FIXME` to the latter!

Reply:
I overlooked the difference in generic params on the scalar `arg`, oops